lightning/src/ln/outbound_payment.rs

```diff
@@ -896,6 +954,119 @@ impl OutboundPayments {
 		Ok(())
 	}
 
+	#[cfg(async_payments)]
+	pub(super) fn static_invoice_received<ES: Deref>(
+		&self, invoice: &StaticInvoice, payment_id: PaymentId, features: Bolt12InvoiceFeatures,
+		best_block_height: u32, entropy_source: ES,
+		pending_events: &Mutex<VecDeque<(events::Event, Option<EventCompletionAction>)>>
+	) -> Result<[u8; 32], Bolt12PaymentError> where ES::Target: EntropySource {
+		macro_rules! abandon_with_entry {
+			($payment: expr, $reason: expr) => {
+				$payment.get_mut().mark_abandoned($reason);
+				if let PendingOutboundPayment::Abandoned { reason, .. } = $payment.get() {
+					if $payment.get().remaining_parts() == 0 {
+						pending_events.lock().unwrap().push_back((events::Event::PaymentFailed {
+							payment_id,
+							payment_hash: None,
+							reason: *reason,
+						}, None));
+						$payment.remove();
+					}
+				}
+			}
+		}
+
+		match self.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+			hash_map::Entry::Occupied(mut entry) => match entry.get() {
+				PendingOutboundPayment::AwaitingInvoice { retry_strategy, invoice_request, .. } => {
+					let invreq = invoice_request.as_ref().ok_or(Bolt12PaymentError::UnexpectedInvoice)?;
+					if !invoice.from_same_offer(invreq) {
+						return Err(Bolt12PaymentError::UnexpectedInvoice)
+					}
+					if invoice.invoice_features().requires_unknown_bits_from(&features) {
+						abandon_with_entry!(entry, PaymentFailureReason::UnknownRequiredFeatures);
+						return Err(Bolt12PaymentError::UnknownRequiredFeatures)
+					}
+					let amount_msat = match InvoiceBuilder::<DerivedSigningPubkey>::amount_msats(&invreq) {
+						Ok(amt) => amt,
+						Err(_) => {
+							debug_assert!(false, "LDK requires an invreq amount for amount-less offers");
```

Comment:
We check this here:
https://github.com/lightningdevkit/rust-lightning/blob/7038906890cd006021eab47601a3c23a3503bac2/lightning/src/offers/invoice_request.rs#L1158-L1163
but I ended up removing this assert since it seems potentially fragile.

Reply:
Ok, I re-added it and added a comment for where we check this.